### PR TITLE
Improving efficiency of large-scale training

### PR DESCRIPTION
**Pull request by pyf98**
## What?

This PR implements several techniques to improve the (memory) efficiency of large-scale training.

- **[Major change]** Implement efficient dataset creation, where only samples belonging to this process are loaded. In distributed training, each process only keeps a fraction of samples, thus significantly reducing memory usage. With this technique, we usually no longer need to split the training set. Splitting data and using multiple iterators can cause 20% to 30% overhead in total training time.
  - Note: `num_workers` also affects the memory usage significantly. Do not use too many workers for large datasets.
- Fix interface compatibility of the latest versions of `flash-attn`
- Enable `amp` for validation
- Support `gradient_as_bucket_view` and `comm_hook` in DDP for better efficiency
  - To use them, add arguments to the training config file: 
    - `gradient_as_bucket_view: true`
    - `ddp_comm_hook: bf16_compress_hook` (for `bf16`)
- Support gradient/activation checkpointing in encoder/decoder for better efficiency
  - In `encoder_conf`, add `gradient_checkpoint_layers: [0]     # 0 is embedding layer, e-branchformer layers starting from 1`
  - In `decoder_conf`, add `gradient_checkpoint_layers: [1, 2, 3, 4, ]  # starting from 1`
  - NOTE: DeepSpeed or Lightning may have their ways of enabling gradient checkpointing. We don't need to set these arguments in ESPnet in those cases.

### CODE REVIEW

**Comment by sw005320** [comment]
Thanks, @pyf98!
Please fix the CI error https://github.com/espnet/espnet/actions/runs/12900093420/job/35969949432?pr=6024

@wanchichen, can you review this PR?

**Comment by pyf98** [comment]
OWSM training works well using this code.

I'm launching a standard ASR job with a lightning trainer to see if there is any issue.

File: espnet2/asr/decoder/transformer_decoder.py [hunk 1]
```diff
@@ -147,9 +153,14 @@ def forward(
         x = self.embed(tgt)
         intermediate_outs = []
         for layer_idx, decoder_layer in enumerate(self.decoders):
-            x, tgt_mask, memory, memory_mask = decoder_layer(
-                x, tgt_mask, memory, memory_mask
-            )
+            if layer_idx + 1 in self.gradient_checkpoint_layers:
```
[the remark this answers is not in the document]
Comment (pyf98) [reply espnet2/asr/decoder/transformer_decoder.py:156]: OK. Lightning may also have another way of doing activation checkpointing. In those customized trainers, we can just disable checkpointing in ESPnet. It should have no conflict.

File: espnet2/train/trainer.py [hunk 2]
```diff
@@ -268,7 +270,29 @@ def run(
                         else None
                     ),
                     find_unused_parameters=trainer_options.unused_parameters,
+                    gradient_as_bucket_view=trainer_options.gradient_as_bucket_view,
                 )
+
+                # Register DDP communication hook
+                if trainer_options.ddp_comm_hook is not None:
+                    from torch.distributed.algorithms.ddp_comm_hooks.default_hooks import (  # noqa: E501
+                        bf16_compress_hook,
```
[the remark this answers is not in the document]
Comment (pyf98) [reply espnet2/train/trainer.py:279]: My opinion is that we don't need. PyTorch 1.10 is too old and we should not maintain it anymore. I actually even want to remove the other very old version checks (but I didn't do it).

[the remark this answers is not in the document]
Comment (pyf98) [reply espnet2/train/trainer.py:279]: By default, this is not used. So, I think it won't affect old versions as well.

File: espnet2/asr/decoder/transformer_decoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pyf98) [reply espnet2/asr/decoder/transformer_decoder.py:156]: By default, this is not used. So, I think it won't conflict with any others.

File: espnet2/train/trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sw005320) [reply espnet2/train/trainer.py:279]: Yes, PyTorch 1.10 is too old, and we don't have to care about it.

**Comment by sw005320** [comment]
I want to make sure whether some core changes in attention went through our current tests (I think so, but I want to make sure).

Other than that, LGTM.
I'll merge after the confirmation about the test.

**Comment by sw005320** [comment]
> I want to make sure whether some core changes in attention went through our current tests (I think so, but I want to make sure).
> 

@pyf98, can you answer it?

**Comment by pyf98** [comment]
> I want to make sure whether some core changes in attention went through our current tests (I think so, but I want to make sure).
> 
> Other than that, LGTM. I'll merge after the confirmation about the test.

The changes in attention are minor and do not affect the normal attention implementation. So I think this part is fine.

**Comment by sw005320** [comment]
Yes, I understand it. It is also going through our test anyway.
